File: server/go/internal/storage/interface.go
// Package storage provides database adapters for document persistence.
// Matches TypeScript reference: server/typescript/src/storage/interface.ts
package storage

import (
	"context"
	"time"
)

// DocumentState represents a stored document
type DocumentState struct {
	ID        string                 `json:"id"`
	State     map[string]interface{} `json:"state"`
	Version   int64                  `json:"version"`
	CreatedAt time.Time              `json:"createdAt"`
	UpdatedAt time.Time              `json:"updatedAt"`
}

// VectorClockEntry represents a vector clock entry for a document
type VectorClockEntry struct {
	DocumentID string    `json:"documentId"`
	ClientID   string    `json:"clientId"`
	ClockValue int64     `json:"clockValue"`
	UpdatedAt  time.Time `json:"updatedAt"`
}

// DeltaEntry represents an operation in the audit trail
type DeltaEntry struct {
	ID            string                 `json:"id"`
	DocumentID    string                 `json:"documentId"`
	ClientID      string                 `json:"clientId"`
	OperationType string                 `json:"operationType"` // "set", "delete", "merge"
	FieldPath     string                 `json:"fieldPath"`
	Value         map[string]interface{} `json:"value,omitempty"`
	ClockValue    int64                  `json:"clockValue"`
	Timestamp     time.Time              `json:"timestamp"`
}

// SessionEntry represents an active connection session
type SessionEntry struct {
	ID          string                 `json:"id"`
	UserID      string                 `json:"userId"`
	ClientID    string                 `json:"clientId,omitempty"`
	ConnectedAt time.Time              `json:"connectedAt"`
	LastSeen    time.Time              `json:"lastSeen"`
	Metadata    map[string]interface{} `json:"metadata,omitempty"`
}

// SnapshotEntry represents a point-in-time document snapshot
type SnapshotEntry struct {
	ID         string                 `json:"id"`
	DocumentID string                 `json:"documentId"`
	State      map[string]interface{} `json:"state"`
	Version    map[string]int64       `json:"version"` // Vector clock at snapshot
	SizeBytes  int                    `json:"sizeBytes"`
	Compressed bool                   `json:"compressed"`
	CreatedAt  time.Time              `json:"createdAt"`
}

// TextDocumentState represents a SyncText (Fugue CRDT) document
type TextDocumentState struct {
	ID        string    `json:"id"`
	Content   string    `json:"content"`   // Plain text content
	CRDTState string    `json:"crdtState"` // Full Fugue CRDT JSON state
	Clock     int64     `json:"clock"`     // Lamport clock
	CreatedAt time.Time `json:"createdAt"`
	UpdatedAt time.Time `json:"updatedAt"`
}

// CleanupOptions specifies what to clean up
type CleanupOptions struct {
	OldSessionsHours        int
	OldDeltasDays           int
	OldSnapshotsDays        int
	MaxSnapshotsPerDocument int
}

// CleanupResult contains cleanup statistics
type CleanupResult struct {
	SessionsDeleted  int `json:"sessionsDeleted"`
	DeltasDeleted    int `json:"deltasDeleted"`
	SnapshotsDeleted int `json:"snapshotsDeleted"`
}

// StorageAdapter defines the interface for document persistence
type StorageAdapter interface {
	// Connection lifecycle
	Connect(ctx context.Context) error
	Disconnect(ctx context.Context) error
	IsConnected() bool
	HealthCheck(ctx context.Context) (bool, error)

	// Document operations
	GetDocument(ctx context.Context, id string) (*DocumentState, error)
	SaveDocument(ctx context.Context, id string, state map[string]interface{}) (*DocumentState, error)
	UpdateDocument(ctx context.Context, id string, state map[string]interface{}) (*DocumentState, error)
	DeleteDocument(ctx context.Context, id string) (bool, error)
	ListDocuments(ctx context.Context, limit, offset int) ([]*DocumentState, error)

	// Vector clock operations
	GetVectorClock(ctx context.Context, documentID string) (map[string]int64, error)
	UpdateVectorClock(ctx context.Context, documentID, clientID string, clockValue int64) error
	MergeVectorClock(ctx context.Context, documentID string, clock map[string]int64) error

	// Delta operations (for audit trail)
	SaveDelta(ctx context.Context, delta *DeltaEntry) (*DeltaEntry, error)
	GetDeltas(ctx context.Context, documentID string, limit int) ([]*DeltaEntry, error)

	// Session operations (for connection tracking)
	SaveSession(ctx context.Context, session *SessionEntry) (*SessionEntry, error)
	UpdateSession(ctx context.Context, sessionID string, lastSeen time.Time, metadata map[string]interface{}) error
	DeleteSession(ctx context.Context, sessionID string) (bool, error)
	GetSessions(ctx context.Context, userID string) ([]*SessionEntry, error)

	// Snapshot operations
	SaveSnapshot(ctx context.Context, snapshot *SnapshotEntry) (*SnapshotEntry, error)
	GetSnapshot(ctx context.Context, snapshotID string) (*SnapshotEntry, error)
	GetLatestSnapshot(ctx context.Context, documentID string) (*SnapshotEntry, error)
	ListSnapshots(ctx context.Context, documentID string, limit int) ([]*SnapshotEntry, error)
	DeleteSnapshot(ctx context.Context, snapshotID string) (bool, error)

	// Text document operations (for SyncText/Fugue CRDT)
	SaveTextDocument(ctx context.Context, id, content, crdtState string, clock int64) (*TextDocumentState, error)
	GetTextDocument(ctx context.Context, id string) (*TextDocumentState, error)

	// Maintenance
	Cleanup(ctx context.Context, options *CleanupOptions) (*CleanupResult, error)
}

// StorageConfig holds configuration for storage adapters
type StorageConfig struct {
	ConnectionString  string
	PoolMinConns      int32
	PoolMaxConns      int32
	ConnectionTimeout time.Duration
}

// DefaultStorageConfig returns sensible defaults
func DefaultStorageConfig() *StorageConfig {
	return &StorageConfig{
		PoolMinConns:      2,
		PoolMaxConns:      10,
		ConnectionTimeout: 5 * time.Second,
	}
}

File: server/python/src/synckit_server/storage/interface.py
"""Storage interface definitions matching TypeScript reference"""

from abc import ABC, abstractmethod
from dataclasses import dataclass, field
from datetime import datetime
from typing import Any, Optional


@dataclass
class DocumentState:
    """Document state with metadata"""
    id: str
    state: Any
    version: int
    created_at: datetime
    updated_at: datetime


@dataclass
class VectorClockEntry:
    """Vector clock entry for a document/client pair"""
    document_id: str
    client_id: str
    clock_value: int
    updated_at: datetime


@dataclass
class DeltaEntry:
    """Delta operation for audit trail"""
    id: str
    document_id: str
    client_id: str
    operation_type: str  # 'set', 'delete', 'merge'
    field_path: str
    value: Optional[Any]
    clock_value: int
    timestamp: datetime


@dataclass
class SessionEntry:
    """Session tracking entry"""
    id: str
    user_id: str
    client_id: Optional[str]
    connected_at: datetime
    last_seen: datetime
    metadata: dict = field(default_factory=dict)


@dataclass
class SnapshotEntry:
    """Document state snapshot"""
    id: str
    document_id: str
    state: Any
    version: dict[str, int]  # Vector clock at time of snapshot
    size_bytes: int
    created_at: datetime
    compressed: bool = False


@dataclass
class TextDocumentState:
    """Text document state for SyncText/Fugue CRDT"""
    id: str
    content: str
    crdt_state: str
    clock: int
    created_at: datetime
    updated_at: datetime


@dataclass
class CleanupOptions:
    """Options for cleanup operation"""
    old_sessions_hours: int = 24
    old_deltas_days: int = 30
    old_snapshots_days: int = 7
    max_snapshots_per_document: int = 10


@dataclass
class CleanupResult:
    """Result of cleanup operation"""
    sessions_deleted: int
    deltas_deleted: int
    snapshots_deleted: int


class StorageAdapter(ABC):
    """Abstract storage adapter interface"""

    # Connection lifecycle
    @abstractmethod
    async def connect(self) -> None:
        pass

    @abstractmethod
    async def disconnect(self) -> None:
        pass

    @abstractmethod
    def is_connected(self) -> bool:
        pass

    @abstractmethod
    async def health_check(self) -> bool:
        pass

    # Document operations
    @abstractmethod
    async def get_document(self, id: str) -> Optional[DocumentState]:
        pass

    @abstractmethod
    async def save_document(self, id: str, state: Any) -> DocumentState:
        pass

    @abstractmethod
    async def update_document(self, id: str, state: Any) -> DocumentState:
        pass

    @abstractmethod
    async def delete_document(self, id: str) -> bool:
        pass

    @abstractmethod
    async def list_documents(self, limit: int = 100, offset: int = 0) -> list[DocumentState]:
        pass

    # Vector clock operations
    @abstractmethod
    async def get_vector_clock(self, document_id: str) -> dict[str, int]:
        pass

    @abstractmethod
    async def update_vector_clock(self, document_id: str, client_id: str, clock_value: int) -> None:
        pass

    @abstractmethod
    async def merge_vector_clock(self, document_id: str, clock: dict[str, int]) -> None:
        pass

    # Delta operations (audit trail)
    @abstractmethod
    async def save_delta(self, delta: DeltaEntry) -> DeltaEntry:
        pass

    @abstractmethod
    async def get_deltas(self, document_id: str, limit: int = 100) -> list[DeltaEntry]:
        pass

    # Session operations
    @abstractmethod
    async def save_session(self, session: SessionEntry) -> SessionEntry:
        pass

    @abstractmethod
    async def update_session(self, session_id: str, last_seen: datetime, metadata: Optional[dict] = None) -> None:
        pass

    @abstractmethod
    async def delete_session(self, session_id: str) -> bool:
        pass

    @abstractmethod
    async def get_sessions(self, user_id: str) -> list[SessionEntry]:
        pass

    # Snapshot operations
    @abstractmethod
    async def save_snapshot(self, snapshot: SnapshotEntry) -> SnapshotEntry:
        pass

    @abstractmethod
    async def get_snapshot(self, snapshot_id: str) -> Optional[SnapshotEntry]:
        pass

    @abstractmethod
    async def get_latest_snapshot(self, document_id: str) -> Optional[SnapshotEntry]:
        pass

    @abstractmethod
    async def list_snapshots(self, document_id: str, limit: int = 10) -> list[SnapshotEntry]:
        pass

    @abstractmethod
    async def delete_snapshot(self, snapshot_id: str) -> bool:
        pass

    # Text document operations (SyncText/Fugue CRDT)
    @abstractmethod
    async def save_text_document(
        self, id: str, content: str, crdt_state: str, clock: int
    ) -> TextDocumentState:
        pass

    @abstractmethod
    async def get_text_document(self, id: str) -> Optional[TextDocumentState]:
        pass

    # Maintenance
    @abstractmethod
    async def cleanup(self, options: Optional[CleanupOptions] = None) -> CleanupResult:
        pass
